### DISCUSSION TITLE
Restoring IDs

### DISCUSSION THREAD

**Discussion by alexturpin** [hmans/miniplex#314] [Q&A]
Hey all! I want to keep Multiplex worlds on different devices in sync; I saw the documentation on how to get IDs from entities, but it's not clear how I can create an entity with that same ID on the other client. Is there something I'm missing?

Thank you!

**Comment by hmans** [comment]
Hi Alex,

the multiplayer/network sync story of Miniplex is pretty much unexplored still (at least from my end), and the existing ID functionality is mostly meant as a quick helper to allow people who need entity IDs to get them.

I do understand your use case, though, and would like to add this functionality so you can optionally pass an ID to the `add` function to immediately register the given entity under that ID.

Unfortunately I don't have a lot of time to work on Miniplex these days (partly due to me currently not building anything with it myself), but I'd be happy to accept a PR for this if anyone wants to have a stab at it.

**Reply by alexturpin** [reply]
Thanks for the reply! Totally understand. I'm excited about it for a current project so would be more than happy to contribute.

**Reply by alexturpin** [reply]
One of the problems with that approach is unfortunately the sequential numerical IDs; it's not really reliable to figure out where to start from if entities have yet to be synced and there could be conflicts of IDs if not enforced on the server.

Entities with unique random IDs (e.g. paralleldrive/cuid2) are better suited for this, but changing the types of IDs might be a bit too much if others use numerical. I can always build a map with string IDs similar to how the current ones work.

**Reply by hmans** [reply]
Yeah, I think for the time being your best option is to develop your own ID management system that maps IDs to entities and back using weakmaps, or similar. The ID system that's built into Miniplex right now is mostly supposed to be a quick and simple helper, and I've always felt that it might even be the best option to just remove it, and instead give examples for implementing custom ID systems (and maybe publishing one as a separate package.)

I'll keep this in mind for when I next get the chance to work on this library.

**Reply by alexturpin** [reply]
That might be the move, thanks for the feedback!

**Comment by cgarvis** [comment]
@alexturpin what did you end up doing?
